pandas/core/generic.py

```diff
@@ -8857,16 +8851,18 @@ def _align_series(
             right = right.fillna(fill_value, method=method, limit=limit)
 
         # if DatetimeIndex have different tz, convert to UTC
-        if is_series or (not is_series and axis == 0):
-            if is_datetime64tz_dtype(left.index.dtype):
-                if left.index.tz != right.index.tz:
-                    if join_index is not None:
-                        # GH#33671 ensure we don't change the index on
-                        #  our original Series (NB: by default deep=False)
-                        left = left.copy()
-                        right = right.copy()
-                        left.index = join_index
-                        right.index = join_index
+        if (
+            (is_series or axis == 0)
+            and is_datetime64tz_dtype(left.index.dtype)
+            and left.index.tz != right.index.tz
+            and join_index is not None
+        ):
+            # GH#33671 ensure we don't change the index on
+            #  our original Series (NB: by default deep=False)
+            left = left.copy()
+            right = right.copy()
+            left.index = join_index
+            right.index = join_index
```

Comment:
identical(?) to chunk above, could make a helper.  also im pretty sure mypy is going to complain about Index not having a tz attribute

Reply:
@jbrockmendel do you know when this code is meant to be reached? Running
```
pytest pandas/tests/generic/
```
doesn't cover it. I couldn't get here with `test_fast.sh` either, even though [codecov shows it as covered](https://codecov.io/gh/pandas-dev/pandas/src/master/pandas/core/generic.py)

Reply:
Looks like `test_alignment_doesnt_change_tz` from #36503 hits the other usage of this pattern, might not have gotten both